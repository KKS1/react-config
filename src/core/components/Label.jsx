import React, { Component } from "react";

const Label = ({ schema, label, id }) => {
  return !schema && label ? <label htmlFor={id}>{label}</label> : "";
};

export default Label;
