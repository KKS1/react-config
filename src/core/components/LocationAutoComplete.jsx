import React, { Component } from "react";
import Label from "./Label";
/* global google */

const ref = React.createRef();

export default class LocationAutoComplete extends Component {
  static defaultProps = {
    placeholder: "",
    onChange: () => {},
    onPlaceChange: () => {}
  };

  constructor(props) {
    super(props);
    this.autocomplete = null;
    this.hasFocused = false;
  }

  componentDidMount() {
    this.hasFocused = false;

    if (!$.R.GooglePlaces.autocomplete_init) {
      $.R.GooglePlaces.autocompleteFocus();
    }
  }

  componentWillUnmount() {
    this.hasFocused = false;
  }

  focusHandler = () => {
    if (this.hasFocused) return;

    this.hasFocused = true;

    this.autocomplete = new google.maps.places.Autocomplete(ref.current, {
      types: ["geocode"]
    });

    this.autocomplete.addListener("place_changed", this.handlePlaceChanged);
    $.R.GooglePlaces.reappendGoogleContainer(ref.current.parentElement, 0);
  };

  handlePlaceChanged = () => {
    const place = this.autocomplete.getPlace();
    this.props.onPlaceChange(place);
  };

  changeHandler = e => {
    this.props.onChange(e.target.value);
  };

  render() {
    const { placeholder, value, options, id, name } = this.props;
    return (
      <div className="c-label-field">
        <Label for={id} {...this.props} />
        <input
          ref={ref}
          id={id}
          placeholder={placeholder}
          type="text"
          className="address"
          aria-haspopup="true"
          role="combobox"
          aria-autocomplete="both"
          aria-label="Type your text here to display the list of address choices."
          name={name}
          value={value == null ? "" : value}
          onFocus={this.focusHandler}
          onChange={this.changeHandler}
        />
      </div>
    );
  }
}
