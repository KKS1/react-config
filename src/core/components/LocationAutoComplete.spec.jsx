/* global google */
import React, { Component } from "react";
import Label from "./Label";
import LocationAutoComplete from "./LocationAutoComplete";

beforeAll(() => {
  $.R = {
    GooglePlaces: {
      autocomplete_init: false,
      autocompleteFocus: jest.fn(),
      reappendGoogleContainer: jest.fn()
    }
  };
});

describe("LocationAutoComplete", () => {
  test("render_LocationAutoComplete", () => {
    const props = {
      id: "formOne_labelOne"
    };
    const wrapper = shallow(<LocationAutoComplete {...props} />);
    const instance = wrapper.instance();
    expect(instance).toMatchSnapshot();
  });

  test("call_onChange_prop", () => {
    const changeHandler = jest.fn();

    $.R.GooglePlaces.autocomplete_init = true;

    const props = {
      id: "formOne_labelOne",
      value: "303 2nd st",
      onChange: changeHandler
    };
    const wrapper = shallow(<LocationAutoComplete {...props} />);
    const instance = wrapper.instance();
    expect(instance).toMatchSnapshot();

    instance.changeHandler({ target: { value: "304 3rd St" } });

    expect(changeHandler).toHaveBeenCalledWith("304 3rd St");
  });

  test("call_onPlaceChange_prop", () => {
    const placeChangeHandler = jest.fn();

    const props = {
      id: "formOne_labelOne",
      value: "303 2nd St",
      onPlaceChange: placeChangeHandler
    };
    const wrapper = shallow(<LocationAutoComplete {...props} />);
    const instance = wrapper.instance();

    instance.autocomplete = jest.fn(() => {
      return {
        getPlace: () => ({
          formatted_address: "303 2nd St, San Francisco, CA 94107, USA"
        })
      };
    })();
    expect(instance).toMatchSnapshot();

    instance.handlePlaceChanged();

    expect(placeChangeHandler).toHaveBeenCalledWith({
      formatted_address: "303 2nd St, San Francisco, CA 94107, USA"
    });
  });

  test("focusHandler", () => {
    const props = {
      id: "formOne_labelOne",
      value: "303 2nd St"
    };
    const wrapper = mount(<LocationAutoComplete {...props} />);
    const instance = wrapper.instance();

    google = jest.fn(() => ({
      maps: {
        places: {
          Autocomplete: () => ({
            addListener: jest.fn()
          })
        }
      }
    }))();

    instance.focusHandler();

    expect(instance.autocomplete.addListener).toHaveBeenCalled();
    expect($.R.GooglePlaces.reappendGoogleContainer).toHaveBeenCalled();

    instance.focusHandler();

    wrapper.unmount();
  });
});
