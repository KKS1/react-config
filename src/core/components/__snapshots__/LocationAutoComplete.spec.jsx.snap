// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationAutoComplete call_onChange_prop 1`] = `
LocationAutoComplete {
  "autocomplete": null,
  "changeHandler": [Function],
  "context": Object {},
  "focusHandler": [Function],
  "handlePlaceChanged": [Function],
  "hasFocused": false,
  "props": Object {
    "id": "formOne_labelOne",
    "onChange": [MockFunction],
    "onPlaceChange": [Function],
    "placeholder": "",
    "value": "303 2nd st",
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useState": [Function],
      },
      "_element": <LocationAutoComplete
        id="formOne_labelOne"
        onChange={[MockFunction]}
        onPlaceChange={[Function]}
        placeholder=""
        value="303 2nd st"
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <div
        className="c-label-field"
      >
        <Label
          for="formOne_labelOne"
          id="formOne_labelOne"
          onChange={[MockFunction]}
          onPlaceChange={[Function]}
          placeholder=""
          value="303 2nd st"
        />
        <input
          aria-autocomplete="both"
          aria-haspopup="true"
          aria-label="Type your text here to display the list of address choices."
          className="address"
          id="formOne_labelOne"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          role="combobox"
          type="text"
          value="303 2nd st"
        />
      </div>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;

exports[`LocationAutoComplete call_onPlaceChange_prop 1`] = `
LocationAutoComplete {
  "autocomplete": Object {
    "getPlace": [Function],
  },
  "changeHandler": [Function],
  "context": Object {},
  "focusHandler": [Function],
  "handlePlaceChanged": [Function],
  "hasFocused": false,
  "props": Object {
    "id": "formOne_labelOne",
    "onChange": [Function],
    "onPlaceChange": [MockFunction],
    "placeholder": "",
    "value": "303 2nd St",
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useState": [Function],
      },
      "_element": <LocationAutoComplete
        id="formOne_labelOne"
        onChange={[Function]}
        onPlaceChange={[MockFunction]}
        placeholder=""
        value="303 2nd St"
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <div
        className="c-label-field"
      >
        <Label
          for="formOne_labelOne"
          id="formOne_labelOne"
          onChange={[Function]}
          onPlaceChange={[MockFunction]}
          placeholder=""
          value="303 2nd St"
        />
        <input
          aria-autocomplete="both"
          aria-haspopup="true"
          aria-label="Type your text here to display the list of address choices."
          className="address"
          id="formOne_labelOne"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          role="combobox"
          type="text"
          value="303 2nd St"
        />
      </div>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;

exports[`LocationAutoComplete render_LocationAutoComplete 1`] = `
LocationAutoComplete {
  "autocomplete": null,
  "changeHandler": [Function],
  "context": Object {},
  "focusHandler": [Function],
  "handlePlaceChanged": [Function],
  "hasFocused": false,
  "props": Object {
    "id": "formOne_labelOne",
    "onChange": [Function],
    "onPlaceChange": [Function],
    "placeholder": "",
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useState": [Function],
      },
      "_element": <LocationAutoComplete
        id="formOne_labelOne"
        onChange={[Function]}
        onPlaceChange={[Function]}
        placeholder=""
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <div
        className="c-label-field"
      >
        <Label
          for="formOne_labelOne"
          id="formOne_labelOne"
          onChange={[Function]}
          onPlaceChange={[Function]}
          placeholder=""
        />
        <input
          aria-autocomplete="both"
          aria-haspopup="true"
          aria-label="Type your text here to display the list of address choices."
          className="address"
          id="formOne_labelOne"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          role="combobox"
          type="text"
          value=""
        />
      </div>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;
